src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,147 @@
+
+import com.amazonaws.secretsmanager.caching.SecretCache;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
+import com.amazonaws.services.secretsmanager.model.ResourceNotFoundException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+public class AWSSecretsManagerPlugin implements IConnectionPlugin {
+  private static final String SECRET_ID_PROPERTY = "secretsManagerSecretId";
+  private static final String REGION_PROPERTY = "secretsManagerRegion";
+  private static final String ERROR_MISSING_DEPENDENCY =
+      "[AWSSecretsManagerPlugin] Required dependencies 'AWS Java SDK for AWS Secrets Manager' and/or 'AWS Secrets Manager Java Caching Client' are not on the classpath";
+  private static final String ERROR_GET_SECRETS_FAILED =
+      "[AWSSecretsManagerPlugin] Was not able to either fetch or read the database credentials from AWS Secrets Manager. Ensure the correct secretId and region properties have been provided";
+  private static final String ERROR_INITIAL_CONN_FAILED =
+      "[AWSSecretsManagerPlugin] Open initial connection failed with given credentials. Retrieving secrets was not attempted";
+  private static final String SQLSTATE_ACCESS_ERROR = "28000";
+
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final String secretId;
+  private final SecretCache secretCache;
+
+  public AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) throws SQLException {
+
+    try {
+      Class.forName("com.amazonaws.services.secretsmanager.AWSSecretsManager");
+      Class.forName("com.amazonaws.secretsmanager.caching.SecretCache");
+    } catch (ClassNotFoundException e) {
+      logger.logError(ERROR_MISSING_DEPENDENCY);
+      throw new SQLException(ERROR_MISSING_DEPENDENCY);
+    }
+
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.secretId = propertySet.getStringProperty(SECRET_ID_PROPERTY).getValue();
+    final String region = propertySet.getStringProperty(REGION_PROPERTY).getValue();
+    this.secretCache = new SecretCache(AWSSecretsManagerClientBuilder.standard().withRegion(region).build());
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    try {
+      // Attempt to open initial connection with provided credentials
+      this.nextPlugin.openInitialConnection(connectionUrl);
+
+    } catch (SQLException connectionFailedException) {
+
+      // If the thrown error is because user access was denied, get new credentials
+      if (Objects.equals(connectionFailedException.getSQLState(), SQLSTATE_ACCESS_ERROR)) {
```

Comment:
Why not to use a simple "=="?

Reply:
https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java
Objects.equals is preferred when comparing strings in Java

Reply:
I see. Since SQLSTATE_ACCESS_ERROR is constant and it's never null, more readable code would be
SQLSTATE_ACCESS_ERROR.equals(connectionFailedException.getSQLState())